### PR TITLE
Fix for "name your form" link

### PR DESCRIPTION
**Pull request by daine**
# Summary | Résumé

This PR fixes issue #2115 
- #2115

_For any UI changes please include screenshots at different breakpoints. It would also be helpful to include a
before and after comparison for each change i.e._

**New look** The placeholder for the filename "Unnamed Form FIle" used to be black. Not sure if this was intentional, so I changed it to look the same as the form title wherein it's grayed out until the user starts typing.

| Before                                          | After                                        |
| ----------------------------------------------- | -------------------------------------------- |
| Screenshot 2023-06-05 at 3 26 23 PM| Screenshot 2023-06-05 at 3 27 57 PM
 |Input on focus:Screenshot 2023-06-05 at 3 26 27 PM|Input on focus:Screenshot 2023-06-05 at 3 28 06 PM|

# Test instructions | Instructions pour tester la modification

Using the form builder, leave the file name blank. Attempt to share the form by hitting the "Share" button at the top right. You will receive an error message saying "You must name your form". Click on the "name your form" link. It should make the cursor focus on the text input for naming the form, which has a placeholder "Unnamed form file"

# Unresolved questions / Out of scope | Questions non résolues ou hors sujet

- [x] Need to ask designers if changing the placeholder colour from black to gray is approved @samsadasivan @anikbrazeau @connorscarolyns 

# Pull Request Checklist

Please complete the following items in the checklist before you request a review:

- [x] Have you completely tested the functionality of change introduced in this PR? Is the PR solving the problem it's meant to solve within the scope of the related issue?
- [x] The PR does not introduce any new issues such as failed tests, console warnings or new bugs.
- [ ] Is the code clean, readable and maintainable? Is it easy to understand and comprehend.
- [ ] Does your code have adequate comprehensible comments? Do new functions have [docstrings](https://tsdoc.org/)?
- [ ] Have you modified the change log and updated any relevant documentation?
- [ ] Is there adequate test coverage? Both unit tests and end-to-end tests where applicable?
- [ ] If your PR is touching any UI is it accessible? Have you tested it with a screen reader? Have you tested it with automated testing tools such as axe?

### CODE REVIEW

**Review by craigzour — DISMISSED** [review]
Works well!
I see one unresolved question where you would like the approval of design team. Will approve and let you make sure you talk to them before merging ;)

**Review by bryan-robitaille — CHANGES_REQUESTED** [review]
I'll try to provide some technical guidance on how we can move away from driving component state through url params after lunch but I noticed 2 things in testing.

1. If the form name is not completed after focus a user can still share the form.
save with no form name
2. If the form name is not completed and a user shifts focus elsewhere on the page the focus button will not return focus to the form name input a second time.
share state resets

**Comment by craigzour** [comment]
> I'll try to provide some technical guidance on how we can move away from driving component state through url params after lunch but I noticed 2 things in testing.
> 
> 1. If the form name is not completed after focus a user can still share the form.
>    save with no form name
> 2. If the form name is not completed and a user shifts focus elsewhere on the page the focus button will not return focus to the form name input a second time.
>    share state resets

I think the reason why you can share even if the form name is empty is because you have put something in the form title. If I remember correctly we needed either one to be able to share a form.

**Comment by daine** [comment]
@bryan-robitaille Thank you! Some thoughts:
> I'll try to provide some technical guidance on how we can move away from driving component state through url params after lunch but I noticed 2 things in testing.

Thanks! The changes in this PR follows the pattern and implementation of the old code so I can see why it needs to be updated

> > 1. If the form name is not completed after focus a user can still share the form.

> I think the reason why you can share even if the form name is empty is because you have put something in the form title. If I remember correctly we needed either one to be able to share a form. Good catch on the second issue!

I believe @craigzour is correct, I'm working on a separate (but related) issue to address the issues with form names and titles

> 2. If the form name is not completed and a user shifts focus elsewhere on the page the focus button will not return focus to the form name input a second time.

This is the same behaviour with the old one that I've found, so it's not a new bug but rather a leftover from the old behavior. I think it's definitely tied in with potentially just changing the previous implementation altogether!

**Comment by samsadasivan** [comment]
Hey Daine! I like this - could we just switch the shade to gray-500 (#6b7280) for colour contrast for AA. slate-500 (#64748b) also works but may be too fancy hehe 


image

**Comment by timarney** [comment]
> Let's work together to get this into a more 'React' state of mind.

@bryan-robitaille fyi if your thinking about refs it gets a little tricky given where the components get built. For the "most" of the form builder elements now have a refs provider which store the "els" in an array.   The "file name" is part of the header hard to get the ref over to the share modal --- the form builder also does quite a bit of focusing so if I remember correctly it was "stealing focus"  Perhaps you have some tricks for that sort of thing :)

**Comment by timarney** [comment]
As far as the code goes this fixes the issue as described.

Before approving ... from the design end do we want to allow sharing an empty form that has been given a filename?

Related ....

I can't find an issue for it but the "file name" used to auto-populate when a "file title was filled out".    It does fill it out if you enter a "form title" and than navigate to another page.  Fixing that could help.

**Comment by samsadasivan** [comment]
It definitely used to auto-populated in FR/EN

**Review by craigzour — APPROVED** [review]
LGTM :) Nice work!

File: components/form-builder/app/navigation/FileName.tsx [hunk 1]
```diff
@@ -17,6 +19,33 @@ export const FileNameInput = () => {
...
+
+  // React Hook that lets you customize the handle exposed as a ref
+  // In this case we are only exposing the focus function
+  useImperativeHandle(
+    remoteRef,
+    () => {
+      return {
+        focus() {
+          fileNameInput.current?.focus();
+          logMessage.debug("Calling focus function on fileNameInput");
```
Comment (timarney) [inline components/form-builder/app/navigation/FileName.tsx:42]: Let's remove the debug log here.

**Review by craigzour — DISMISSED** [review]

File: lib/hooks/useRefStore.tsx [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+const RefStoreContext = createContext<RefStore | undefined>(undefined);
+
+export const RefStoreProvider = ({ children }: { children: React.ReactNode }) => {
+  const refStore = useRef(new Map<string, React.RefObject<HTMLElement>>());
+
+  const getRef = useCallback(
+    (key: string) => {
+      const value = refStore.current.get(key);
+      logMessage.debug(`getRef :key: ${key}, value: ${value}`);
```
Comment (timarney) [inline lib/hooks/useRefStore.tsx:17]: Probably don't need this anymore?

File: lib/hooks/useRefStore.tsx [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+      return value;
+    },
+    [refStore]
+  );
+
+  const setRef = useCallback(
+    (key: string, ref: React.RefObject<HTMLElement>) => {
+      const store = refStore.current;
+      store.set(key, ref);
+      logMessage.debug(`setRef :key: ${key}, ref: ${ref}`);
```
Comment (timarney) [inline lib/hooks/useRefStore.tsx:27]: Probably don't need this anymore?

**Review by timarney — DISMISSED** [review]
Approving --- just a couple of debug messages left

File: components/form-builder/app/navigation/FileName.tsx [hunk 4]
```diff
@@ -17,6 +18,32 @@ export const FileNameInput = () => {
...
+    setRef("fileNameInput", remoteRef);
+
+    return () => {
+      removeRef("fileNameInput");
+    };
+  }, [remoteRef, setRef, removeRef]);
+
+  // React Hook that lets you customize the handle exposed as a ref
+  // In this case we are only exposing the focus function
+  useImperativeHandle(
```
Comment (thiessenp-cds) [inline components/form-builder/app/navigation/FileName.tsx:35]: Nice!

**Review by thiessenp-cds — APPROVED** [review]
I took a quick look at the accessibility issue with the Share menu. The sub menu will need to add some keyboard navigation etc. 

I can confirm that this issue is an existing issue (develop branch). I think it makes sense to open a new issue about this, out of scope etc. I'll approve the work along side other approvals. Nice work! :)

**Pull request merged**